tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11457,6 +11499,34 @@ class WorkGraphVisitor : public RecursiveASTVisitor<WorkGraphVisitor> {
     }
     return true;
   }
+
+  bool VisitAttr(Attr *A) {
+    switch (A->getKind()) {
+    case attr::HLSLNodeDispatchGrid:
+    case attr::HLSLNodeMaxDispatchGrid:
+      // These attributes are only valid for Broadcasting launch nodes
```

Comment:
In general I don’t think we should use an AST visitor to validate these. We can instead just look at the attributes on the Decl after we complete attribute parsing.

Reply:
For now I have moved these checks up to the FunctionDecl visitor - as a (small) step toward consolidating all function attribute checks into a single verification function.

Reply:
The WorkGraphVisitor has now been removed, and the diagnostics raised there have all been relocated.